pkg/clients/aws/route53.go

```diff
@@ -611,16 +611,17 @@ func getSTSCredentials(reqLogger logr.Logger, client *sts.STS, roleArn string, e
 		if err == nil {
 			break
 		}
-		if i == (assumeRolePollingRetries - 1) {
-			return nil, fmt.Errorf("timed out while assuming role %s", roleArn)
+		// Error was returned
+		if i >= (assumeRolePollingRetries - 1) {
+			// If we're out of retries, return the wrapped error
+			return nil, fmt.Errorf("failed to assumeRole, no retries left: %w", err)
 		}
-	}
-	if err != nil {
 		// Log AWS error
 		if aerr, ok := err.(awserr.Error); ok {
```

Comment:
This isn't going to work anymore, as it depends on the err variable that's not removed. This will need to be folded into the for loop now, probably on line 614. Then you can switch on the specific AWS error and either exit immediately, or do a retry

Reply:
The err has been mentioned in L607

Also, we don't really need to cast it as an awserr and remove the whole check mentioned below?

```
			// Log AWS error
			if aerr, ok := err.(awserr.Error); ok {
				reqLogger.Error(aerr, "New AWS Error while getting STS credentials,\nAWS Error Code: %s,\nAWS Error Message: %s", aerr.Code(), aerr.Message())
			}
```

Reply:
Okay, I think I see now that this is handled within the for loop. It still important to cast it as an AWS error IMO to differentiate it from other errors, and as mentioned, depending on what the specific error is from AWS, you could exit the for loop early because retrying would be pointless.

Reply:
I assume we are already casting the errors as aws error.  You mean we shall return out the AWS like below?
```
		// Log AWS error
		if aerr, ok := err.(awserr.Error); ok {
			return nil, fmt.Errorf("\"%w\", New AWS Error while getting STS credentials,\nAWS Error Code: %s,\nAWS Error Message: %s", aerr, aerr.Code(), aerr.Message())
		}
```

OR, as you said depending on specific error from AWS, exit loop early so, only way we can achieve it, is by the case statement shown below?

`		if aerr, ok := err.(awserr.Error); ok {
			reqLogger.Error(aerr, "AWS Error while assuming role - Code: %s, Message: %s", aerr.Code(), aerr.Message())

			// Handle specific AWS errors
			switch aerr.Code() {
			case sts.ErrCodeThrottlingException, sts.ErrCodeLimitExceededException:
				// For throttling or limit exceeded errors, continue retrying
				reqLogger.Warn("Retryable AWS error encountered: %s", aerr.Code())
				continue

			case sts.ErrCodeExpiredTokenException:
				// For expired token errors, stop retrying as it indicates an issue with credentials
				return nil, fmt.Errorf("expired token error: %w", aerr)

			default:
				// Handle other AWS errors as needed
				reqLogger.Error("Unhandled AWS error: %s", aerr.Code())
				return nil, fmt.Errorf("AWS error: %w", aerr)
			}`

Reply:
I'd venture a guess as to the second case.

Just the only change I'd make to the logic therein would be having the default case attempt to retry so we don't always _immediately_ exit out.

Reply:
Kirk got it :)